### ISSUE TITLE
Changing project name breaks all C# scripts

### ISSUE THREAD

**Issue by mateusak** [godotengine/godot#27030]
**Godot version:**
Godot 3.1

**OS/device including version:**
Windows 10

**Issue description:**
Changing Project Settings > Application > Config > Name and running the project gives the error "Cannot instance script because the assembly project is not loaded" to all C# scripts. If you close and open Godot again after this, however, it fixes the issue.

**Comment by asmoaesl** [comment]
Agree that this happened to me, as well.

**Comment by ScepticDope** [comment]
Anyone found another way to fix a project? Just restarting the editor does not work for me.

**Comment by ScepticDope** [comment]
Found a fix with some help from @ForLoveOfCats. To rename a project just manually change the name everywhere in your project files. I missed `Properties/AssemblyInfo.cs` when I tried that earlier.

**Comment by ScepticDope** [comment]
Found a temporary fix with some help from @ForLoveOfCats. To rename a project just manually change the name everywhere in your project files. Make sure to check `Properties/AssemblyInfo.cs`.

**Comment by NathanWarden** [comment]
I'm still of the opinion that, at least for now, the generated C# project and related files/properties should just be called GodotProject (or something like that). This would eliminate the issue of a project rename breaking everything since no matter what the project is named it will always be able to find GodotProject.sln and its related files.

**Comment by neikeq** [comment]
My idea was to create an project setting for the csproj name that is filled with the project name the first time it's created, but is not updated if the project name changes.

**Comment by akien-mga** [comment]
Is this still reproducible in latest 3.2 builds?

**Comment by akien-mga** [comment]
Still reproducible in 3.2 RC 1 Mono. E.g. after renaming the "DodgeTheCreepsCS" demo to "DodgeTheCreeps CS", pressing "Run" starts the game fine but scripts are broken:
```
ERROR: can_instance: Cannot instance script because the project assembly is not loaded. Script: 'res://Main.cs'.
   At: modules/mono/csharp_script.cpp:2892.
ERROR: can_instance: Cannot instance script because the project assembly is not loaded. Script: 'res://Player.cs'.
   At: modules/mono/csharp_script.cpp:2892.
ERROR: can_instance: Cannot instance script because the project assembly is not loaded. Script: 'res://HUD.cs'.
   At: modules/mono/csharp_script.cpp:2892.
ERROR: emit_signal: Error calling method from signal 'pressed': 'CanvasLayer(HUD.cs)::OnStartButtonPressed': Method not found..
   At: core/object.cpp:1228
```

As mentioned above though, restarting the editor fixes it, so it's not critical to fix for the 3.2 release.

**Comment by tuckersn** [comment]
How I fixed this issue was manually editing "old_project_name.sln" and "old_project_name.csproj".
1. "old_project_name.sln" and "old_project_name.csproj" need to be renamed to  "new_project_name.sln" and "new_project_name.csproj".
2. Open the .csproj file. Change "RootNamespace" to the new name with the spaces removed if you have any, and change "AssemblyName" to the name with the spaces kept.
3. Open the .sln and change the "old_project_name" on line 3 to "new_project_name" (there are two places you need to make this change). 

**I believe this is case-sensitive, although I haven't tested it.**

I believe to fix this when a project is renamed in the mono build these values need to be modified, the csproj would be easy if Godot has an XML parser.

Tested on v3.2.2 mono.

**Comment by tuckersn** [comment]
> How I fixed this issue was manually editing "old_project_name.sln" and "old_project_name.csproj".
> 
> 1. "old_project_name.sln" and "old_project_name.csproj" need to be renamed to  "new_project_name.sln" and "new_project_name.csproj".
> 2. Open the .csproj file. Change "RootNamespace" to the new name with the spaces removed if you have any, and change "AssemblyName" to the new name with the spaces kept.
> 3. Open the .sln and change the "old_project_name" on line 3 to "new_project_name" (there are two places you need to make this change).
> 
> **I believe this is case-sensitive, although I haven't tested it.**
> 
> I believe to fix this when a project is renamed in the mono build these values need to be modified, the csproj would be easy if Godot has an XML parser.
> 
> Tested on v3.2.2 mono.



> How I fixed this issue was manually editing "old_project_name.sln" and "old_project_name.csproj".
> 
> 1. "old_project_name.sln" and "old_project_name.csproj" need to be renamed to  "new_project_name.sln" and "new_project_name.csproj".
> 2. Open the .csproj file. Change "RootNamespace" to the new name with the spaces removed if you have any, and change "AssemblyName" to the name with the spaces kept.
> 3. Open the .sln and change the "old_project_name" on line 3 to "new_project_name" (there are two places you need to make this change).
> 
> **I believe this is case-sensitive, although I haven't tested it.**
> 
> I believe to fix this when a project is renamed in the mono build these values need to be modified, the csproj would be easy if Godot has an XML parser.
> 
> Tested on v3.2.2 mono.

**Comment by nbourre** [comment]
Following @tuckersn comment, I used Visual Studio Code to replace all occurence of the old name with the new name. 

Use [Ctrl] + [Shift] + F to find all occurence of the old name and replace all of them.

**Comment by KevinTriplett** [comment]
When changing project name, it would be nice to also change the root folder name. I see there are many files that reference the old project name, so I think it's not simple to implement (I'm new so my perspective is limited).

**Comment by tommai78101** [comment]
We could perhaps implement something akin to Unreal Engine, where you can run a simple batch executable file that would re-generate the project solution files, but tweak it so that it is using the information found in the actual, crucial files that Godot requires to see.

https://github.com/wallerm/UE4/blob/master/Engine/Build/BatchFiles/GenerateProjectFiles.bat

**Comment by Domigome** [comment]
> How I fixed this issue was manually editing "old_project_name.sln" and "old_project_name.csproj".
> 
> 1. "old_project_name.sln" and "old_project_name.csproj" need to be renamed to  "new_project_name.sln" and "new_project_name.csproj".
> 2. Open the .csproj file. Change "RootNamespace" to the new name with the spaces removed if you have any, and change "AssemblyName" to the new name with the spaces kept.
> 3. Open the .sln and change the "old_project_name" on line 3 to "new_project_name" (there are two places you need to make this change).
> 
> **I believe this is case-sensitive, although I haven't tested it.**
> 
> I believe to fix this when a project is renamed in the mono build these values need to be modified. Tested on v3.2.2 mono.

Great, I just broke a project after finishing it because of willing to change the final name people will see.

Hope a fix for this could be implemented or at least, LOCK the project name editing option in the editor.

**Comment by andoriann** [comment]
It's been 3 years. Any updates for when it will be fixed? Still reproduces in 3.4.5.

**Comment by Calinou** [comment]
> Any updates for when it will be fixed?

There is no ETA for fixing issues, as contributors work on a best-effort basis.

**Comment by neikeq** [comment]
This will be solved in master/4.0 once #64089 is merged. For 3.x we need to backport 9c698629499a9a8fffb74a3ec4c2bfac6b7a1484. Things have changed a lot, so it won't be a simple cherry-pick. Unfortunately, I'm entirely focused on the 4.0 milestone right now, so I don't have time.

**Comment by raulsntos** [comment]
Opened #64460 to backport 9c698629499a9a8fffb74a3ec4c2bfac6b7a1484 to 3.x when the commit lands in master.

**Comment by akien-mga** [comment]
Fixed by #64089 (4.0) and #64460 (3.6).
